cluster.go

```diff
@@ -459,13 +471,42 @@ func (c *Cluster) Do(a Action) error {
 	return c.doInner(a, addr, key, false, doAttempts)
 }
 
+func (c *Cluster) isClusterDown() bool {
+	return atomic.LoadUint64(&c.down) == 1
+}
+
+func (c *Cluster) setClusterDown(down bool) bool {
+	var changed bool
+	if down {
+		changed = atomic.CompareAndSwapUint64(&c.down, 0, 1)
+	} else {
+		changed = atomic.CompareAndSwapUint64(&c.down, 1, 0)
+	}
+
+	if changed && c.co.ct.StateChange != nil {
+		c.co.ct.StateChange(trace.ClusterStateChange{IsDown: down})
+	}
+
+	return changed
+}
+
 func (c *Cluster) traceRedirected(addr, key string, moved, ask bool, attempts int) {
 	if c.co.ct.Redirected != nil {
-		c.co.ct.Redirected(trace.ClusterRedirected{Addr: addr, Key: key, Moved: moved, Ask: ask, RedirectCount: doAttempts - attempts})
+		c.co.ct.Redirected(trace.ClusterRedirected{
+			Addr:          addr,
+			Key:           key,
+			Moved:         moved,
+			Ask:           ask,
+			RedirectCount: doAttempts - attempts,
+		})
 	}
 }
 
 func (c *Cluster) doInner(a Action, addr, key string, ask bool, attempts int) error {
+	if c.co.clusterDownWait > 0 && c.isClusterDown() {
+		time.Sleep(c.co.clusterDownWait)
```

Comment:
Again, simplest thing to implement for prototyping. In my test this works really great and avoids complex synchronization logic with other goroutines.